### PR TITLE
feat(watcher): add support for poll watcher

### PR DESCRIPTION
**Pull request by amribm**
Change for adding poll watcher for vector config change 

Closes https://github.com/vectordotdev/vector/issues/21111

### CODE REVIEW

File: src/config/watcher.rs [hunk 1]
```diff
@@ -19,21 +19,60 @@ const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(1
 
 const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
 
+pub enum ConfigWatcher {
```
Comment (jorgehermo9) [inline src/config/watcher.rs:22]: I found confusing that this PR includes two `ConfigWatcher` struct with the same name but different meanings

File: src/config/watcher.rs [hunk 2]
```diff
@@ -90,23 +130,23 @@ pub fn spawn_thread<'a>(
...
+//     Error,
+// > {
+//     info!("Creating configuration file watcher.");
+//     let (sender, receiver) = channel();
+//     let mut watcher = recommended_watcher(sender)?;
+//     add_paths(&mut watcher, config_paths)?;
+//     Ok((watcher, receiver))
+// }
+
+fn add_paths(watcher: &mut Box<dyn Watcher>, config_paths: &[PathBuf]) -> Result<(), Error> {
```
Comment (jorgehermo9) [inline src/config/watcher.rs:149]: I think you could use
```suggestion
fn add_paths<T: Watcher>(watcher: &mut T, config_paths: &[PathBuf]) -> Result<(), Error> {
```

**Comment by jorgehermo9** [comment]
Just as a recommendation, if you start your pr title with `Closes #<issue-number>`, will link this PR to the issue and close the issue when the PR is closed

File: src/config/watcher.rs (hunk 1 above)
Comment (amribm) [reply src/config/watcher.rs:23]: yeah , one i used  enum for clap options and another one in watcher.

if u can suggest me a different name. i will change

File: src/cli.rs [hunk 3]
```diff
@@ -152,6 +154,23 @@ pub struct RootOpts {
     #[arg(short, long, env = "VECTOR_WATCH_CONFIG")]
     pub watch_config: bool,
 
+    /// Method for watching config
+    ///
+    /// By default `vector` use `inotify` which is recommended for linux based system
+    /// `poll` watcher can be used where inotify not work. ie, attaching config using NFS
+    #[arg(long, default_value = "inotify", env = "VECTOR_WATCH_METHOD")]
```
Comment (jorgehermo9) [inline src/cli.rs:161]: the `ConfigWatcher::Inotify` variant does not always set the `inotify` watcher, it sets the `RecommendedWatcher`, which would be `inotify` on supported platforms and polling in others.

I think we should rename those variants (or at least, the `Inotify` one), as it is a bit misleading for the user (also, teh default value here should be change accordingly)

File: src/cli.rs [hunk 4]
```diff
@@ -354,10 +373,23 @@ pub enum LogFormat {
...
     for error in errors {
         error!(message = "Configuration error.", %error);
     }
 
     exitcode::CONFIG
 }
+
+pub fn watcher_config(method: ConfigWatcher, interval: NonZeroU64) -> watcher::ConfigWatcher {
+    match method {
+        ConfigWatcher::Inotify => watcher::ConfigWatcher::RecommendedWatcher,
```
Comment (jorgehermo9) [inline src/cli.rs:392]: the `ConfigWatcher::Inotify` variant does not always set the `inotify` watcher, it sets the `RecommendedWatcher`, which would be `inotify` on supported platforms and polling in others.

I think we should rename those variants (or at least, the `Inotify` one), as it is a bit misleading for the user (also, the default value in the args should be change accordingly)

**Comment by amribm** [comment]
i had some commits without my github email and CLA is not passing. in effort to ammend the username in commit . i messed up the tree. when i force pushed this PR got closed. 

Though i already created new one. https://github.com/vectordotdev/vector/pull/21290

**Pull request closed without merging**